cql3/query_processor.cc

```diff
@@ -732,62 +732,30 @@ bool query_processor::has_more_results(::shared_ptr<cql3::internal_query_state>
-        is_done = seastar::make_shared<bool>(false);
-
-        auto stop_when = [is_done]() {
-            return *is_done;
-        };
-        auto do_resuls = [is_done, state, f, this]() mutable {
-            return this->execute_paged_internal(
-                    state).then([is_done, state, f, this](::shared_ptr<cql3::untyped_result_set> msg) mutable {
-                if (msg->empty()) {
-                    *is_done = true;
-                } else {
-                    if (!this->has_more_results(state)) {
-                        *is_done = true;
-                    }
-                    for (auto& row : *msg) {
-                        if (f(row) == stop_iteration::yes) {
-                            *is_done = true;
-                            break;
-                        }
-                    }
-                }
-            });
-        };
-        return do_until(stop_when, do_resuls);
-    });
+    do {
+        auto msg = co_await execute_paged_internal(state);
+        if (msg->empty()) {
+            co_return;
+        }
+        for (auto& row : *msg) {
+            if (f(row) == stop_iteration::yes) {
+                co_return;
+            }
+        }
+    } while (has_more_results(state));
 }
 
 future<> query_processor::for_each_cql_result(
         ::shared_ptr<cql3::internal_query_state> state,
          noncopyable_function<future<stop_iteration>(const cql3::untyped_result_set::row&)>&& f) {
-    // repeat can move the lambda's capture, so we need to hold f and it so the internal loop
-    // will be able to use it.
-    return do_with(noncopyable_function<future<stop_iteration>(const cql3::untyped_result_set::row&)>(std::move(f)),
-            untyped_result_set::rows_type::const_iterator(),
-            [state, this](noncopyable_function<future<stop_iteration>(const cql3::untyped_result_set::row&)>& f,
-                    untyped_result_set::rows_type::const_iterator& it) mutable {
-        return repeat([state, &f, &it, this]() mutable {
-            return this->execute_paged_internal(state).then([state, &f, &it, this](::shared_ptr<cql3::untyped_result_set> msg) mutable {
-                it = msg->begin();
-                return repeat_until_value([&it, &f, msg, state, this]() mutable {
-                    if (it == msg->end()) {
-                        return make_ready_future<std::optional<stop_iteration>>(std::optional<stop_iteration>(!this->has_more_results(state)));
-                    }
-
-                    return f(*it).then([&it, msg](stop_iteration i) {
-                        if (i == stop_iteration::yes) {
-                            return std::optional<stop_iteration>(i);
-                        }
-                        ++it;
-                        return std::optional<stop_iteration>();
-                    });
-                });
-            });
-        });
-    });
+    do {
+        auto msg = co_await execute_paged_internal(state);
```

Comment:
We either need the `msg->empty()` check here too, or we don't need it above.

Reply:
I think it's okay. It's a direct translation and coroutinization shouldn't try to perform extra optimizations.

As it happens I think the empty check is a bug. Suppose we receive an empty page but has_more_pages = true? But it should be fixed separately.

Reply:
There's also no need to have these two overloads. The only difference is returning future<stop_iteration> vs. non-future. The callers can be adjusted, or one can call the other.

Reply:
I will fix both in a follow up pr with two different commits.

Reply:
It's better to avoid mixing fixes and refactoring.

Reply:
sure. will create a follow-up PR once this one is merged.

Reply:
> sure. will create a follow-up PR once this one is merged.

#13675